core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,246 @@
+use super::{
+    hasher::{self, Digest},
+    AdviceSetError, Word,
+};
+use math::fields::f64::BaseElement;
+use std::collections::HashMap;
+
+// MERKLE PATH SET
+// ================================================================================================
+
+/// A map of Merkle Paths
+///
+/// This struct is intended to be used as one of the variants of the MerkleSet enum.
+#[derive(Clone, Debug)]
+pub struct MerklePathSet {
+    root: Word,
+    total_depth: u32,
+    paths: HashMap<u64, Vec<Word>>,
+}
+
+impl MerklePathSet {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an empty MerklePathSet.
+    pub fn new(depth: u32) -> Result<Self, AdviceSetError> {
+        let root = [BaseElement::new(0); 4];
+        let paths = HashMap::<u64, Vec<Word>>::new();
+
+        Ok(Self {
+            root,
+            total_depth: depth,
+            paths,
+        })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds the Merkle Path to the paths map.
+    pub fn add_path(
+        &mut self,
+        depth: u32,
+        index: u64,
+        path: Vec<Word>,
+    ) -> Result<(), AdviceSetError> {
+        let index = 2u64.pow(depth) + index / 2;
+        self.paths.insert(index, path);
+        Ok(())
+    }
```

Comment:
For the first version let's simplify this: we can have paths only of the same length. So, we don't need the `depth` parameter.

Also, we can probably get rid of `2u64.pow(depth)` part since basing the map key just on the index should be enough.

On the other hand, we should probably check if the index and the path length are consistent with the depth of the Merkle path set.